### PR TITLE
Modin implementation of DataFrame API standard

### PR DESCRIPTION
**Pull request by anmyachev**
@MarcoGorelli thanks for the notice in https://github.com/data-apis/dataframe-api-compat/pull/69#issuecomment-1908692910!

This pull request is the result of an unfinished integration that I have stopped at for now.

Perhaps I can continue to work on some things that are needed for integration, but do not directly related to Modin.

## **Updates 4/24/2024:**

**Main changes:**
* Modin implementation with `ModinHandler` for tests (mostly copied from `pandas_standard`). 
* Tests for modin are run using `--library modin` flag and in a separate environment: `requirements-dev-modin.txt`. By default (without `--library` parameter) they do not start.
* Mark tests with `xfail` or `skip` marks through `pytest_collection_modifyitems` (as it done in https://github.com/data-apis/dataframe-interchange-tests/blob/main/tests/conftest.py)
* Lazy import submodules (`pandas_standard`, `polars_standard`, `modin_standard`) via `__getattr__`. This is mainly due to the fact that `mypy` followed import statement for `polars_standard`  and checked the hint type when a folder for another implementation was specified (`dataframe_api_compat/modin_standard`), and since at that moment (in CI) polars had already been deleted, this led to errors. The same problem occurred when running the check as follows `mypy tests`. However, here this problem has been solved by using `--follow-imports silent` flag.
* The code has been rewritten so that tests for implementations other than polars can be run even if it is not installed (in the future, the same thing will most likely need to be done for pandas dependency).

### CODE REVIEW

File: tests/conftest.py [hunk 1]
```diff
@@ -2,19 +2,120 @@
...
+        return (
+            dataframe_api_compat.modin_standard.convert_to_standard_compliant_dataframe(
+                df,
+                api_version=api_version or "2023.11-beta",
+            )
+        )
+
+
+@pytest.fixture(scope="session", params=LIBRARIES[sys.version_info[:2]])
+def library(request):
```
[the remark this answers is not in the document]
Comment (MarcoGorelli) [reply tests/conftest.py:112]: sounds good!

**Comment by MarcoGorelli** [comment]
hi @anmyachev - just FYI, I've asked to be removed from the project

maybe we can work something out together if you're interested, but it needs to be outside of the consortium

**Comment by anmyachev** [comment]
> hi @anmyachev - just FYI, I've asked to be removed from the project

Hi @MarcoGorelli, thanks for letting me know. I'm sad that this happened. Although I have only been involved in the protocol for a short time, let me express my gratitude for your enormous contribution to its development.

> maybe we can work something out together if you're interested, but it needs to be outside of the consortium

Also, taking into account your contributions to open source in general, I believe our work will intersect.

File: .github/workflows/earliest_versions.yml [hunk 2]
```diff
@@ -31,4 +31,4 @@ jobs:
       - name: install-reqs
         run: python -m pip install --upgrade tox virtualenv setuptools pip && python -m pip install -r requirements-dev.txt
       - name: Run pytest
-        run: pytest tests --cov=dataframe_api_compat --cov=tests --cov-fail-under=80
+        run: pytest tests --cov=dataframe_api_compat/pandas_standard --cov=dataframe_api_compat/polars_standard --cov=tests --cov-fail-under=80
```
Comment (anmyachev) [inline .github/workflows/earliest_versions.yml:34]: To avoid using modin code for measurement since it is executed in a separate CI job.

File: .github/workflows/mkdocs.yml [hunk 3]
```diff
@@ -27,6 +27,6 @@ jobs:
           path: .cache
           restore-keys: |
             mkdocs-material-
-      - run: pip install -r docs/requirements-docs.txt -e . pandas polars
+      - run: pip install -r docs/requirements-docs.txt -e . pandas polars # modin[ray] @ git+https://github.com/modin-project/modin.git@abbbd03023f55d4ee9592c4c695308e10a97cb4c
```
Comment (anmyachev) [inline .github/workflows/mkdocs.yml:30]: The main branch has fixes that are needed for some tests.

File: .github/workflows/tox.yml [hunk 4]
```diff
@@ -29,7 +29,7 @@ jobs:
       - name: install-reqs
         run: python -m pip install --upgrade tox virtualenv setuptools pip -r requirements-dev.txt
       - name: Run pytest
-        run: pytest tests --cov=dataframe_api_compat --cov=tests --cov-fail-under=100
+        run: pytest tests --cov=dataframe_api_compat/pandas_standard --cov=dataframe_api_compat/polars_standard --cov=tests --cov-fail-under=98
```
Comment (anmyachev) [inline .github/workflows/tox.yml:32]: There is code in the `tests` folder that only Modin uses, so there cannot be 100 percent coverage.

File: .github/workflows/tox.yml [hunk 5]
```diff
@@ -39,6 +39,35 @@ jobs:
       - name: run pandas integration tests
         run: pip uninstall polars -y && pytest tests/integration/upstream_test.py::TestPandas
 
+  tox-modin:
```
Comment (anmyachev) [inline .github/workflows/tox.yml:42]: Modin's environment may not always be compatible with pandas and polars, so a separate job is required (perhaps it is unnecessary).

File: pyproject.toml [hunk 6]
```diff
@@ -82,6 +82,12 @@ line-length = 90
 filterwarnings = [
   "error",
   'ignore:distutils Version classes are deprecated:DeprecationWarning',
+  'ignore:.*pkg_resources:DeprecationWarning',
+  'ignore:Ray execution environment not yet initialized:UserWarning',
+  "ignore:Distributing <class '.*'> object.:UserWarning",
+  'ignore:.*ray:ResourceWarning',
+  'ignore:.*is not currently supported by PandasOnRay:UserWarning',
+  'ignore:.*implementation has mismatches with pandas:UserWarning',
```
Comment (anmyachev) [inline pyproject.toml:90]: Looks like it possible to setup these filters in pytest runtime (if needed).

File: tests/column/and_or_test.py [hunk 7]
```diff
@@ -1,10 +1,11 @@
 from __future__ import annotations
 
+from tests.utils import BaseHandler
 from tests.utils import bool_dataframe_1
 from tests.utils import compare_column_with_reference
 
 
-def test_column_and(library: str) -> None:
+def test_column_and(library: BaseHandler) -> None:
```
Comment (anmyachev) [inline tests/column/and_or_test.py:8]: I can make this change into a separate pull request if the approach as a whole is satisfactory.

File: tests/utils.py [hunk 8]
```diff
@@ -1,40 +1,182 @@
 from __future__ import annotations
 
 import math
+from abc import abstractmethod
```
Comment (anmyachev) [inline tests/utils.py:4]: Most of the changes can be moved into a separate pull request if the approach as a whole is satisfactory.

**Comment by anmyachev** [comment]
Hi @rgommers, I don't know if you have write access to the repository, but perhaps you could review the PR.

**Comment by rgommers** [comment]
> I don't know if you have write access to the repository,

I do. Just checking: you do too, right? If not, we should fix that.

> but perhaps you could review the PR

I could, but not this week - pretty swamped, and this is not a small PR. I think we should revisit this once we have a better plan for the way forward for this package.

**Comment by anmyachev** [comment]
> > I don't know if you have write access to the repository,
> 
> I do. Just checking: you do too, right? If not, we should fix that.

It looks like I haven't (for example, I can't merge the PR myself).

> > but perhaps you could review the PR
> 
> I could, but not this week - pretty swamped, and this is not a small PR. I think we should revisit this once we have a better plan for the way forward for this package.

No problem. Thank you!

**Comment by rgommers** [comment]
> It looks like I haven't (for example, I can't merge the PR myself).

Ah, the `all-members` team was missing from this repo. Does it work now?

**Comment by anmyachev** [comment]
> > It looks like I haven't (for example, I can't merge the PR myself).
> 
> Ah, the `all-members` team was missing from this repo. Does it work now?

Yes, it works. Thanks!

**Comment by anmyachev** [comment]
Hi @rgommers! Since I have write rights to this repository, I can commit without review by the consortium. What do you think about it? At first I would like to just add an implementation for Modin and rewrite the tests to be more independent of the backend where possible.

From a global perspective, I don't plan to do anything that contradicts the statement on the main page of this repository: `Extremely lightweight compatibility layer between pandas, Polars, and any other project which implements the [Python Dataframe API Standard](https://data-apis.org/dataframe-api/).`

**Comment by rgommers** [comment]
Hey @anmyachev, thanks for proposing that. I think that is a very good idea. As long as we're short on reviewer bandwidth, I think it's perfectly fine for you to move ahead (and the same goes for other maintainers). This compat layer can be very light on review.

File: tests/integration/upstream_test.py [hunk 9]
```diff
@@ -7,7 +7,7 @@
 
 class TestPolars:
     def test_dataframe(self) -> None:
-        import polars as pl
+        pl = pytest.importorskip("polars")
```
Comment (anmyachev) [inline tests/integration/upstream_test.py:10]: To avoid using such constructs, may need to reconsider either the location of such tests (tests written for a specific implementation) or the procedure for running them.
